### DISCUSSION TITLE
Temperature Script for cameras

### DISCUSSION THREAD

**Discussion by vastmeridian** [aaronwmorris/indi-allsky#560] [Q&A] [answered]
The camera I am using (Altair 178C) does not have a temperature sensor that is read by the Indi driver.  I am therefore looking at writing a script to record the temperature of the camera on the images.

My question is, how does the script need to report its temperature to allsky?  I don't think I've seen any documentation for this anywhere.

I wrote a very simple script (it just says 'echo "20"') and set it up in the config but I get the result 'Temperature script failed to execute' from the config utility as I try to enter it ( (the script has the execution bit set).

Any ideas?

**Comment by aaronwmorris** [comment, marked as answer]
My documentation is not extensive, but I have some example scripts documented at:
https://github.com/aaronwmorris/indi-allsky/wiki/External-Temperature-Sensor

If you have the exact exception that includes the line number, I might be able to tell you what the problem is.  If you could also run `ls -l script.sh` (or whatever the name is) to show what the file permissions actually are.

**Reply by vastmeridian** [reply]
How did I miss that in the Wiki?  Sorry to have bothered you.  Thanks.

**Reply by vastmeridian** [reply]
To complete the question - I now have simple scripts in both shell and python that read the CPU temperature (it's more accurate than -273.1C) and this now shows in the images.

Before installing the script:
![BeforeScript](https://user-images.githubusercontent.com/15128874/218344439-2b998d95-fd9c-45a9-a11f-946c5fea9c02.jpg)

After installing the script:
![AfterScript](https://user-images.githubusercontent.com/15128874/218344672-b0c352d9-7635-42b1-b6d4-f09b56a3c65f.jpg)

The variable I read is /sys/class/thermal/thermal_zone0/temp

What I don't understand is that when run outside of indi_allsky I get the "Environmental variable not defined" error, but when indi_allsky runs the script there is no error.  Is TEMP_JSON already defined within indi_allsky?

**Reply by aaronwmorris** [reply]
indi-allsky generates the TEMP_JSON environment variable.  It is normally a completely random file name.

You can test it using the following syntax:

```
TEMP_JSON=/tmp/foobar.json ./misc/example_ccd_temp.sh 

cat /tmp/foobar.json
```

After indi-allsky reads the resulting data from the random file name, the file is deleted.

**Comment by pierrebilb** [comment]
Hi,
i'm quite lost with linux and i need help to get a ds18b20 working within indi-allsky.
I tried this script temperature.py:

```
#!/usr/bin/env python3
# Lecture de temperature avec DS18B20

import os
import glob
import time
import sys
import io
import json
import logging

base_dir = '/sys/bus/w1/devices/'
device_folder = glob.glob(base_dir + '28*')[0]
device_file = device_folder + '/w1_slave'

def read_temp_raw():
    f = open(device_file, 'r')
    lines = f.readlines()
    f.close()
    return lines

def read_temp():
    lines = read_temp_raw()
    while lines[0].strip()[-3:] != 'YES':
        time.sleep(0.2)
        lines = read_temp_raw()
    equals_pos = lines[1].find('t=')
    if equals_pos != -1:
        temp_string = lines[1][equals_pos+2:]
        temp_c = float(temp_string) / 1000.0
        return temp_c


# Example of an external temperature script for indi-allsky
# STDOUT and STDERR are ignored
#
# The json output file is set in the environment variable TEMP_JSON


logging.basicConfig(level=logging.INFO)
logger = logging

try:
    # data file is communicated via environment variable
    temp_json = os.environ['TEMP_JSON']
except KeyError:
    logger.error('TEMP_JSON environment variable is not defined')
    sys.exit(1)


# dict to be used for json data
temp_data = {
    'temp' : temp_c,
}


# write json data
with io.open(temp_json, 'w') as f_temp_json:
    json.dump(temp_data, f_temp_json, indent=4)


# script must exist with exit code 0 for success
sys.exit(0)
```

But it doesn't work of course..
I have to modify allsky.py to tell him to run my script?
If someone can guide me..
Thanks a lot
Pierre

**Reply by aaronwmorris** [reply]
You just need to call your function to set the `temp_c` variable.

```python
#!/usr/bin/env python3
# Lecture de temperature avec DS18B20

import os
import glob
import time
import sys
import io
import json
import logging

base_dir = '/sys/bus/w1/devices/'
device_folder = glob.glob(base_dir + '28*')[0]
device_file = device_folder + '/w1_slave'

def read_temp_raw():
    f = open(device_file, 'r')
    lines = f.readlines()
    f.close()
    return lines

def read_temp():
    lines = read_temp_raw()
    while lines[0].strip()[-3:] != 'YES':
        time.sleep(0.2)
        lines = read_temp_raw()
    equals_pos = lines[1].find('t=')
    if equals_pos != -1:
        temp_string = lines[1][equals_pos+2:]
        temp_c = float(temp_string) / 1000.0
        return temp_c


# Example of an external temperature script for indi-allsky
# STDOUT and STDERR are ignored
#
# The json output file is set in the environment variable TEMP_JSON


logging.basicConfig(level=logging.INFO)
logger = logging

try:
    # data file is communicated via environment variable
    temp_json = os.environ['TEMP_JSON']
except KeyError:
    logger.error('TEMP_JSON environment variable is not defined')
    sys.exit(1)

# dict to be used for json data
temp_data = {
    'temp' : read_temp(),  # <--- call function here
}


# write json data
with io.open(temp_json, 'w') as f_temp_json:
    json.dump(temp_data, f_temp_json, indent=4)


# script must exist with exit code 0 for success
sys.exit(0)
```

**Comment by pierrebilb** [comment]
Hi, thanks a lot for your help and congratulation for your program!
i still get
`ERROR:root:TEMP_JSON environment variable is not defined` 
when i run the script inside indi_allsky.
Pierre

**Reply by aaronwmorris** [reply]
If you run your script manually, you have to provide a path to the output file via the TEMP_JSON environment variable.  When indi-allsky runs the script, it will provide the TEMP_JSON path where the script can write this file for every request.

```
TEMP_JSON=/tmp/foobar.json ./script.py
```

**Comment by pierrebilb** [comment]
Hi, thank you very much for your help
i put the script temperature.py (with your mod) in the indi_allsky directory and modify line 83 of config.py: `"CCD_TEMP_SCRIPT"  : "temperature.py"` but it does not work..
I'm sorry to need help again for things so obvious!

**Reply by aaronwmorris** [reply]
You should edit the configuration using the web interface.  It has better error checking.

You may just need to specify the full path to the script.

**Comment by pierrebilb** [comment]
It's working with the web interface with full path!
Thanks a lot for your program and your support!!
👍
